### PR TITLE
lib: at_cmd_parser: AT parser not handling all modem responses correctly

### PR DESCRIPTION
**Pull request by dagbja**
NCSDK-5199: Temporary fix for failing commands (HWVERSION etc.)

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

### CODE REVIEW

**Comment by dagbja** [comment]
This replaces NCSDK-5199, which got too messed up.

**Comment by dagbja** [comment]
All comments from NCSDK-5199 have been fixed, including latest comment about AT+CGSN.

**Comment by coderbyheart** [comment]
> This replaces NCSDK-5199, which got too messed up.

You mean #3379

**Comment by dagbja** [comment]
Yes, #3379. Sorry, mixed up with the Jira number.

**Comment by coderbyheart** [comment]
It would be helpful to include some background information in the commit message to explain what this change addresses because the referenced JIRA issue is only visible to Nordic employees.

**Comment by dagbja** [comment]
Several AT responses from the modem are processed incorrectly, because they are strings without double-quotes. This change will force these to be processed as strings.

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 1]
```diff
@@ -229,6 +266,15 @@ static int at_parse_process_element(const char **str, int index,
 			tmpstr++;
 		}
 
+		at_params_string_put(list, index, start_ptr,
+				     tmpstr - start_ptr);
+	} else if (state == CLAC) {
+		const char *start_ptr = tmpstr;
+
+		while (!is_terminated((int)*tmpstr)) {
```
Comment (rlubos) [inline lib/at_cmd_parser/at_cmd_parser.c:274]: I don't think you need to cast to `int`, this function accepts `char`.

File: lib/at_cmd_parser/at_utils.h [hunk 2]
```diff
@@ -242,6 +242,30 @@ static inline bool is_command(const char *str)
 	return false;
 }
 
+/**
+ * @brief Check if a string is a beginning of an AT CLAC response
+ *
+ * This function will check if the string is a CLAC response prefix.
+ * Valid prefixes: AT+ and AT%, except AT%X
```
Comment (rlubos) [inline lib/at_cmd_parser/at_utils.h:249]: Why is `AT%X` an exception?

File: lib/at_cmd_parser/at_utils.h [hunk 3]
```diff
@@ -242,6 +242,30 @@ static inline bool is_command(const char *str)
...
+ * @param[in] str String to examine
```
Comment (rlubos) [inline lib/at_cmd_parser/at_utils.h:250]: Let's add an empty line above, currently, the param list blends with the description.

File: lib/at_cmd_parser/at_utils.h [hunk 4]
```diff
@@ -242,6 +242,30 @@ static inline bool is_command(const char *str)
...
+	if (strlen(str) < 3) {
+		return false;
+	}
+
+	if ((toupper((int)str[0]) == 'A') &&
+		(toupper((int)str[1]) == 'T') &&
+		((toupper((int)str[2]) == '+') ||
+		((toupper((int)str[2]) == '%') &&  (toupper((int)str[3]) != 'X')))) {
+		return true;
+	}
```
Comment (rlubos) [inline lib/at_cmd_parser/at_utils.h:266]: Perhaps tweak the indentation to make it a bit more readable, something like:
```suggestion
	if ((toupper((int)str[0]) == 'A') &&
	    (toupper((int)str[1]) == 'T') &&
	    ((toupper((int)str[2]) == '+') ||
	     ((toupper((int)str[2]) == '%') &&
	      (toupper((int)str[3]) != 'X')))) {
		return true;
	}
```
Still not the most readable `if` but I don't have a better idea for now.

Comment (coderbyheart) [reply lib/at_cmd_parser/at_utils.h:266]: You could turn the conditions around like this:

```
if (toupper((int)str[0]) != 'A' && toupper((int)str[1]) != 'T') return false; // No AT cmd
if (toupper((int)str[2]) != '+' && toupper((int)str[2]) != '%') return false; // Neither AT+ nor AT%
if (toupper((int)str[2]) == '%' && toupper((int)str[3]) == 'X') return false; // ignore AT%X
return true;
```

Comment (rlubos) [reply lib/at_cmd_parser/at_utils.h:266]: Just note that Zephyr (and thus NCS) coding guideline requires surrounding `{ }` even for single-line blocks, so it'd rather be:
```
if (toupper((int)str[0]) != 'A' && toupper((int)str[1]) != 'T') {
	return false;
}
```
etc.

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 1 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c:274]: Fixed

File: lib/at_cmd_parser/at_utils.h (hunk 2 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_utils.h:249]: There are some responses starting with AT%X that are not CLAC responses: Read response for XCOEX0, XDATAPRFL, XMAGPIO, XOPCONF, and test response for XDEEPSEARCH. They could be falsely processed as CLAC. 
This logic assumes none of these will be the first command listed in the CLAC response, which seems like a good bet.

File: lib/at_cmd_parser/at_utils.h (hunk 3 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_utils.h:251]: Done

File: lib/at_cmd_parser/at_utils.h [hunk 5]
```diff
@@ -242,6 +242,40 @@ static inline bool is_command(const char *str)
...
+		/* Not an AT command */
+		return false;
+	}
+
+	if ((toupper(str[2]) != '+') && (toupper(str[2]) != '%')) {
+		/* Neither AT+ nor AT% */
+		return false;
+	}
+
+	if ((toupper(str[2]) == '%') && (toupper(str[3]) == 'X')) {
```
Comment (lemrey) [inline lib/at_cmd_parser/at_utils.h:272]: `str` is only guaranteed to be at least 3 characters, so `str[3]` might be out-of-bounds, although perhaps not very likely.

File: lib/at_cmd_parser/at_cmd_parser.c [hunk 6]
```diff
@@ -52,6 +64,22 @@ static inline void skip_command_prefix(const char **cmd)
 	(*cmd)++;
 }
 
+static inline bool check_response_for_forced_string(const char *tmpstr)
```
Comment (lemrey) [inline lib/at_cmd_parser/at_cmd_parser.c:67]: If possible, avoid `inline` here. I don't think there is any reason to inline this or `is_clac` down below (other than similar functions in this file being `inline`).

**Review by lemrey — COMMENTED** [review]
Looks good to me. This fix could be documented in the changelog here: `doc/nrf/releases/release-notes-latest.rst`.
Please add a line there, and let's invite a tech writer to review it. /cc @umapraseeda

File: lib/at_cmd_parser/at_utils.h (hunk 5 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_utils.h:272]: Fixed

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 6 above)
Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c:67]: Fixed

File: lib/at_cmd_parser/at_cmd_parser.c (hunk 1 above)
Comment (rlubos) [reply lib/at_cmd_parser/at_cmd_parser.c:274]: The `(int)` cast is still present.

Comment (dagbja) [reply lib/at_cmd_parser/at_cmd_parser.c]: Sorry, missed that one. Fixed.

**Review by rlubos — APPROVED** [review]

**Review by jhn-nordic — APPROVED** [review]

**Review by lemrey — APPROVED** [review]
Thanks

**Pull request merged**